cmd/contacts.go

```diff
@@ -132,21 +204,95 @@ func (x *findContactsCmd) Execute(args []string) error {
 	if x.Username == "" && x.Peer == "" && x.Address == "" {
 		return errMissingSearchInfo
 	}
-	var infos core.ContactInfoQueryResult
-	res, err := executeJsonCmd(POST, "contacts/search", params{
-		opts: map[string]string{
-			"username": x.Username,
-			"peer":     x.Peer,
-			"address":  x.Address,
-			"local":    strconv.FormatBool(x.Local),
-			"limit":    strconv.Itoa(x.Limit),
-			"wait":     strconv.Itoa(x.Wait),
-			"add":      strconv.FormatBool(x.Add),
-		},
-	}, &infos)
-	if err != nil {
-		return err
+
+	resultsCh := make(chan core.ContactQueryResult)
+	outputCh := make(chan interface{})
+
+	cancel := func() {}
+	quit := make(chan os.Signal)
+	signal.Notify(quit, os.Interrupt)
+
+	go func() {
+		defer func() {
+			cancel()
+			close(resultsCh)
+			os.Exit(1)
+		}()
+
+		var res *http.Response
+		var err error
+		res, cancel, err = request(POST, "contacts/search", params{
+			opts: map[string]string{
+				"username": x.Username,
+				"peer":     x.Peer,
+				"address":  x.Address,
+				"local":    strconv.FormatBool(x.Local),
+				"limit":    strconv.Itoa(x.Limit),
+				"wait":     strconv.Itoa(x.Wait),
+			},
+		})
+		if err != nil {
+			outputCh <- err.Error()
+			return
+		}
+		defer res.Body.Close()
+
+		if res.StatusCode >= 400 {
+			body, err := util.UnmarshalString(res.Body)
+			if err != nil {
+				outputCh <- err.Error()
+			} else {
+				outputCh <- body
+			}
+			return
+		}
+
+		decoder := json.NewDecoder(res.Body)
+		for decoder.More() {
+			var result core.ContactQueryResult
+			if err := decoder.Decode(&result); err == io.EOF {
+				return
+			} else if err != nil {
+				outputCh <- err.Error()
+				return
+			}
+			resultsCh <- result
+		}
+	}()
+
+	go func() {
+		for {
+			select {
+			case res, ok := <-resultsCh:
+				if !ok {
+					return
+				}
+
+				data, err := json.MarshalIndent(res, "", "    ")
+				if err == io.EOF {
+					break
+				} else if err != nil {
+					return
+				}
+				outputCh <- string(data)
+			}
+		}
+	}()
+
+	for {
+		select {
+		case val := <-outputCh:
+			output(val)
+
+		case <-quit:
+			fmt.Println("Interrupted")
+			if cancel != nil {
+				fmt.Printf("Canceling...")
+				cancel()
```

Comment:
Allows canceling of the search all the out to the cafes sitting on a search subscription, collecting results.